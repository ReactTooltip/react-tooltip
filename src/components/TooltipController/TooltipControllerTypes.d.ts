import type { CSSProperties } from 'react'

import type {
  PlacesType,
  VariantType,
  WrapperType,
  ChildrenType,
  EventsType,
  PositionStrategy,
  IPosition,
} from 'components/Tooltip/TooltipTypes'

export interface ITooltipController {
  className?: string
  classNameArrow?: string
  content?: string
  html?: string
  place?: PlacesType
  offset?: number
  id?: string
  variant?: VariantType
  anchorId?: string
  wrapper?: WrapperType
  children?: ChildrenType
  events?: EventsType[]
  positionStrategy?: PositionStrategy
  delayShow?: number
  delayHide?: number
  float?: boolean
  noArrow?: boolean
  clickable?: boolean
  style?: CSSProperties
  position?: IPosition
  isOpen?: boolean
  setIsOpen?: (value: boolean) => void
}

declare module 'react' {
  interface HTMLAttributes<T> extends AriaAttributes, DOMAttributes<T> {
    'data-tooltip-place'?: PlacesType
    'data-tooltip-content'?: string
    'data-tooltip-html'?: string
    'data-tooltip-variant'?: VariantType
    'data-tooltip-offset'?: number
    'data-tooltip-wrapper'?: WrapperType
    'data-tooltip-events'?: EventsType[]
    'data-tooltip-position-strategy'?: PositionStrategy
    'data-tooltip-delay-show'?: number
    'data-tooltip-delay-hide'?: number
    'data-tooltip-float'?: boolean
  }
}
