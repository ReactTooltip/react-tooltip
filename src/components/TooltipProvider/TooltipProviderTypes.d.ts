import type { ReactNode, RefObject } from 'react'
import type { ITooltipController } from 'components/TooltipController/TooltipControllerTypes'

export type AnchorRef = RefObject<HTMLElement>

export interface TooltipContextData {
  getTooltipData: (tooltipId?: string) => {
    anchorRefs: Set<AnchorRef>
    activeAnchor: AnchorRef
    attach: (...refs: AnchorRef[]) => void
    detach: (...refs: AnchorRef[]) => void
    setActiveAnchor: (ref: AnchorRef) => void
  }
}

export interface ITooltipWrapper {
  tooltipId?: string
  children: ReactNode
  className?: string

  place?: ITooltipController['place']
  content?: ITooltipController['content']
  html?: ITooltipController['html']
  variant?: ITooltipController['variant']
  offset?: ITooltipController['offset']
  wrapper?: ITooltipController['wrapper']
  events?: ITooltipController['events']
  positionStrategy?: ITooltipController['positionStrategy']
  delayShow?: ITooltipController['delayShow']
  delayHide?: ITooltipController['delayHide']
}
