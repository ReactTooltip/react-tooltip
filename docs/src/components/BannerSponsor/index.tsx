/* eslint-disable import/no-unresolved */
import React from 'react'
// @ts-ignore
import LogoFrigade from '@site/static/img/sponsors/frigade.png'
// @ts-ignore
import LogoDopt from '@site/static/img/sponsors/dopt.png'
import './styles.css'

declare global {
  interface Window {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    dataLayer?: any
  }
}

const SPONSORS = {
  frigade: {
    logo: LogoFrigade,
    title: 'Frigade',
  },
  dopt: {
    logo: LogoDopt,
    title: 'Dopt',
  },
}

interface BannerSponsorProps {
  sponsorKey: keyof typeof SPONSORS
}

const BannerSponsor = ({ sponsorKey }: BannerSponsorProps) => {
  const sponsor = SPONSORS[sponsorKey]

  const onClickSponsorBannerEventHandler = () => {
    if (typeof window !== 'undefined') {
      window.dataLayer = window.dataLayer || []

      window.dataLayer.push({
        event: `click_${sponsorKey}_banner`,
        place: 'sidebar',
      })
    }

    return true
  }

  return (
    <div className="sponsor-banner">
      <a
        href="https://frigade.com/?source=react-tooltip"
        title={sponsor.title}
        target="_blank"
        rel="noreferrer"
        onClick={onClickSponsorBannerEventHandler}
      >
        <img src={sponsor.logo} alt={sponsor.title} />
      </a>
    </div>
  )
}

export default BannerSponsor
