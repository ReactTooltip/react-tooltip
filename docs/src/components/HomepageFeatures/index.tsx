/* eslint-disable import/no-unresolved */
/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable global-require */
import React from 'react'
import clsx from 'clsx'
import styles from './styles.module.css'

type FeatureItem = {
  title: string
  Svg: React.ComponentType<React.ComponentProps<'svg'>>
  description: JSX.Element
}

const FeatureList: FeatureItem[] = [
  {
    title: 'Easy to Use',
    Svg: require('@site/static/img/only-tooltip-top.svg').default,
    description: (
      <>ReactTooltip was designed with love from the ground up to be easily installed and used.</>
    ),
  },
  {
    title: 'Open Source',
    Svg: require('@site/static/img/github.svg').default,
    description: (
      <>
        An Open Source project built by developers to developers, we work together with the
        community to always try to improve ReactTooltip.
      </>
    ),
  },
  {
    title: 'Focus on What Matters',
    Svg: require('@site/static/img/happy-face-tooltip.svg').default,
    description: (
      <>
        Don&rsquo;t waste your time building a tooltip component from scratch, just use ReactTooltip
        and focus on new features or bug fixes.
      </>
    ),
  },
]

function Feature({ title, Svg, description }: FeatureItem) {
  let classVar = styles.featureSvg;
  if(Svg === require('@site/static/img/github.svg').default)
  {
    classVar = styles.featureSvg +" "+ styles.githubLogo;
  }
  return (
    <div className={clsx('col col--4')}>
      <div className="text--center">
        <Svg className={classVar} role="img" />
      </div>
      <div className="text--center padding-horiz--md">
        <h3>{title}</h3>
        <p>{description}</p>
      </div>
    </div>
  )
}

export default function HomepageFeatures(): JSX.Element {
  return (
    <section className={styles.features}>
      <div className="container">
        <div className="row">
          {FeatureList.map((props, idx) => (
            // eslint-disable-next-line react/no-array-index-key
            <Feature key={idx} {...props} />
          ))}
        </div>
      </div>
    </section>
  )
}
