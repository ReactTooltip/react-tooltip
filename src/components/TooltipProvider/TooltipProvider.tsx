import React, {
  createContext,
  PropsWithChildren,
  useCallback,
  useContext,
  useMemo,
  useState,
} from 'react'

import type {
  AnchorRef,
  TooltipContextData,
  TooltipContextDataWrapper,
} from './TooltipProviderTypes'

const DEFAULT_TOOLTIP_ID = 'DEFAULT_TOOLTIP_ID'
const DEFAULT_CONTEXT_DATA: TooltipContextData = {
  anchorRefs: new Set(),
  activeAnchor: { current: null },
  attach: () => {
    /* attach anchor element */
  },
  detach: () => {
    /* detach anchor element */
  },
  setActiveAnchor: () => {
    /* set active anchor */
  },
}

const DEFAULT_CONTEXT_DATA_WRAPPER: TooltipContextDataWrapper = {
  getTooltipData: () => DEFAULT_CONTEXT_DATA,
}

const TooltipContext = createContext<TooltipContextDataWrapper>(DEFAULT_CONTEXT_DATA_WRAPPER)

const TooltipProvider: React.FC<PropsWithChildren> = ({ children }) => {
  const [anchorRefMap, setAnchorRefMap] = useState<Record<string, Set<AnchorRef>>>({
    [DEFAULT_TOOLTIP_ID]: new Set(),
  })
  const [activeAnchorMap, setActiveAnchorMap] = useState<Record<string, AnchorRef>>({
    [DEFAULT_TOOLTIP_ID]: { current: null },
  })

  const attach = (tooltipId: string, ...refs: AnchorRef[]) => {
    setAnchorRefMap((oldMap) => {
      const tooltipRefs = oldMap[tooltipId] ?? new Set()
      refs.forEach((ref) => tooltipRefs.add(ref))
      // create new object to trigger re-render
      return { ...oldMap, [tooltipId]: new Set(tooltipRefs) }
    })
  }

  const detach = (tooltipId: string, ...refs: AnchorRef[]) => {
    setAnchorRefMap((oldMap) => {
      const tooltipRefs = oldMap[tooltipId]
      if (!tooltipRefs) {
        // tooltip not found
        // maybe thow error?
        return oldMap
      }
      refs.forEach((ref) => tooltipRefs.delete(ref))
      // create new object to trigger re-render
      return { ...oldMap }
    })
  }

  const setActiveAnchor = (tooltipId: string, ref: React.RefObject<HTMLElement>) => {
    setActiveAnchorMap((oldMap) => {
      if (oldMap[tooltipId]?.current === ref.current) {
        return oldMap
      }
      // create new object to trigger re-render
      return { ...oldMap, [tooltipId]: ref }
    })
  }

  const getTooltipData = useCallback(
    (tooltipId = DEFAULT_TOOLTIP_ID) => ({
      anchorRefs: anchorRefMap[tooltipId] ?? new Set(),
      activeAnchor: activeAnchorMap[tooltipId] ?? { current: null },
      attach: (...refs: AnchorRef[]) => attach(tooltipId, ...refs),
      detach: (...refs: AnchorRef[]) => detach(tooltipId, ...refs),
      setActiveAnchor: (ref: AnchorRef) => setActiveAnchor(tooltipId, ref),
    }),
    [anchorRefMap, activeAnchorMap, attach, detach],
  )

  const context = useMemo(() => {
    return {
      getTooltipData,
    }
  }, [getTooltipData])

  return <TooltipContext.Provider value={context}>{children}</TooltipContext.Provider>
}

export function useTooltip(tooltipId = DEFAULT_TOOLTIP_ID) {
  return useContext(TooltipContext).getTooltipData(tooltipId)
}

export default TooltipProvider
