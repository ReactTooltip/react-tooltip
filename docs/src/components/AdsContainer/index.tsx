/* eslint-disable import/no-unresolved */
import React from 'react'
import AdsContainerElement from '@site/src/components/AdsContainerElement'
import BannerSponsor from '@site/src/components/BannerSponsor'
import './styles.css'

const AdsContainer = () => {
  return (
    <div className="ads-container fixed">
      <BannerSponsor sponsorKey="frigade" tier="gold" />
      <AdsContainerElement />
      <BannerSponsor sponsorKey="dopt" tier="silver" />
    </div>
  )
}

export default AdsContainer
