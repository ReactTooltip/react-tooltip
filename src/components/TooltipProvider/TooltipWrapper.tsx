import { useEffect, useRef } from 'react'
import classNames from 'classnames'
import { useTooltip } from './TooltipProvider'
import type { ITooltipWrapper } from './TooltipProviderTypes'

const TooltipWrapper = ({
  tooltipId,
  children,
  className,
  place,
  content,
  html,
  variant,
  offset,
  wrapper,
  events,
  positionStrategy,
  delayShow,
  delayHide,
}: ITooltipWrapper) => {
  const { attach, detach } = useTooltip()(tooltipId)
  const anchorRef = useRef<HTMLElement | null>(null)

  useEffect(() => {
    attach(anchorRef)
    return () => {
      detach(anchorRef)
    }
  }, [])

  return (
    <span
      ref={anchorRef}
      className={classNames('react-tooltip-wrapper', className)}
      data-tooltip-place={place}
      data-tooltip-content={content}
      data-tooltip-html={html}
      data-tooltip-variant={variant}
      data-tooltip-offset={offset}
      data-tooltip-wrapper={wrapper}
      data-tooltip-events={events}
      data-tooltip-position-strategy={positionStrategy}
      data-tooltip-delay-show={delayShow}
      data-tooltip-delay-hide={delayHide}
    >
      {children}
    </span>
  )
}

export default TooltipWrapper
