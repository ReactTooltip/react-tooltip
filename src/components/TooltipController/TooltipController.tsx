import { useEffect, useState } from 'react'
import { Tooltip } from 'components/Tooltip'
import type {
  EventsType,
  PositionStrategy,
  PlacesType,
  VariantType,
  WrapperType,
  DataAttribute,
  ITooltip,
} from 'components/Tooltip/TooltipTypes'
import { useTooltip } from 'components/TooltipProvider'
import type { ITooltipController } from './TooltipControllerTypes'

const TooltipController = ({
  id,
  anchorId,
  content,
  html,
  className,
  classNameArrow,
  variant = 'dark',
  place = 'top',
  offset = 10,
  wrapper = 'div',
  children = null,
  events = ['hover'],
  positionStrategy = 'absolute',
  delayShow = 0,
  delayHide = 0,
  float = false,
  noArrow = false,
  clickable = false,
  style,
  position,
  isOpen,
  setIsOpen,
}: ITooltipController) => {
  const [tooltipContent, setTooltipContent] = useState(content || html)
  const [tooltipPlace, setTooltipPlace] = useState(place)
  const [tooltipVariant, setTooltipVariant] = useState(variant)
  const [tooltipOffset, setTooltipOffset] = useState(offset)
  const [tooltipDelayShow, setTooltipDelayShow] = useState(delayShow)
  const [tooltipDelayHide, setTooltipDelayHide] = useState(delayHide)
  const [tooltipFloat, setTooltipFloat] = useState(float)
  const [tooltipWrapper, setTooltipWrapper] = useState<WrapperType>(wrapper)
  const [tooltipEvents, setTooltipEvents] = useState(events)
  const [tooltipPositionStrategy, setTooltipPositionStrategy] = useState(positionStrategy)
  const [isHtmlContent, setIsHtmlContent] = useState(Boolean(html))
  const { anchorRefs, activeAnchor } = useTooltip()(id)

  const getDataAttributesFromAnchorElement = (elementReference: HTMLElement) => {
    const dataAttributes = elementReference?.getAttributeNames().reduce((acc, name) => {
      if (name.startsWith('data-tooltip-')) {
        const parsedAttribute = name.replace(/^data-tooltip-/, '') as DataAttribute
        acc[parsedAttribute] = elementReference?.getAttribute(name) ?? null
      }
      return acc
    }, {} as Record<DataAttribute, string | null>)

    return dataAttributes
  }

  const applyAllDataAttributesFromAnchorElement = (
    dataAttributes: Record<string, string | null>,
  ) => {
    const handleDataAttributes: Record<DataAttribute, (value: string | null) => void> = {
      place: (value) => {
        setTooltipPlace((value as PlacesType) ?? place)
      },
      content: (value) => {
        setTooltipContent(value ?? content)
      },
      html: (value) => {
        setIsHtmlContent(Boolean(value ?? html))
        setTooltipContent(value ?? html ?? content)
      },
      variant: (value) => {
        setTooltipVariant((value as VariantType) ?? variant)
      },
      offset: (value) => {
        setTooltipOffset(value === null ? offset : Number(value))
      },
      wrapper: (value) => {
        setTooltipWrapper((value as WrapperType) ?? 'div')
      },
      events: (value) => {
        const parsed = value?.split(' ') as EventsType[]
        setTooltipEvents(parsed ?? events)
      },
      'position-strategy': (value) => {
        setTooltipPositionStrategy((value as PositionStrategy) ?? positionStrategy)
      },
      'delay-show': (value) => {
        setTooltipDelayShow(value === null ? delayShow : Number(value))
      },
      'delay-hide': (value) => {
        setTooltipDelayHide(value === null ? delayHide : Number(value))
      },
      float: (value) => {
        setTooltipFloat(value === null ? float : Boolean(value))
      },
    }
    // reset unset data attributes to default values
    // without this, data attributes from the last active anchor will still be used
    Object.values(handleDataAttributes).forEach((handler) => handler(null))
    Object.entries(dataAttributes).forEach(([key, value]) => {
      handleDataAttributes[key as DataAttribute]?.(value)
    })
  }

  useEffect(() => {
    setIsHtmlContent(false)
    setTooltipContent(content)
    if (html) {
      // html will take precedence
      setIsHtmlContent(true)
      setTooltipContent(html)
    }
  }, [content, html])

  useEffect(() => {
    const elementRefs = new Set(anchorRefs)

    const anchorById = document.querySelector(`[id='${anchorId}']`) as HTMLElement
    if (anchorById) {
      elementRefs.add({ current: anchorById })
    }

    if (!elementRefs.size) {
      return () => null
    }

    const observerCallback: MutationCallback = (mutationList) => {
      mutationList.forEach((mutation) => {
        if (
          !activeAnchor.current ||
          mutation.type !== 'attributes' ||
          !mutation.attributeName?.startsWith('data-tooltip-')
        ) {
          return
        }
        // make sure to get all set attributes, since all unset attributes are reset
        const dataAttributes = getDataAttributesFromAnchorElement(activeAnchor.current)
        applyAllDataAttributesFromAnchorElement(dataAttributes)
      })
    }

    // Create an observer instance linked to the callback function
    const observer = new MutationObserver(observerCallback)

    // do not check for subtree and childrens, we only want to know attribute changes
    // to stay watching `data-attributes-*` from anchor element
    const observerConfig = { attributes: true, childList: false, subtree: false }

    const element = activeAnchor.current ?? anchorById

    if (element) {
      const dataAttributes = getDataAttributesFromAnchorElement(element)
      applyAllDataAttributesFromAnchorElement(dataAttributes)
      // Start observing the target node for configured mutations
      observer.observe(element, observerConfig)
    }

    return () => {
      // Remove the observer when the tooltip is destroyed
      observer.disconnect()
    }
  }, [anchorRefs, activeAnchor, anchorId])

  const props: ITooltip = {
    id,
    anchorId,
    className,
    classNameArrow,
    content: tooltipContent,
    isHtmlContent,
    place: tooltipPlace,
    variant: tooltipVariant,
    offset: tooltipOffset,
    wrapper: tooltipWrapper,
    events: tooltipEvents,
    positionStrategy: tooltipPositionStrategy,
    delayShow: tooltipDelayShow,
    delayHide: tooltipDelayHide,
    float: tooltipFloat,
    noArrow,
    clickable,
    style,
    position,
    isOpen,
    setIsOpen,
  }

  return children ? <Tooltip {...props}>{children}</Tooltip> : <Tooltip {...props} />
}

export default TooltipController
