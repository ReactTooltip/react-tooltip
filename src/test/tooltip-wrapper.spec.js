import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import '@testing-library/jest-dom'
import { TooltipController as Tooltip } from '../components/TooltipController'
import { TooltipProvider, TooltipWrapper } from '../components/TooltipProvider'

// Tell Jest to mock all timeout functions
jest.useRealTimers()

describe('tooltip props', () => {
  test('basic tooltip component', async () => {
    const { container } = render(
      <TooltipProvider>
        <TooltipWrapper place="bottom" content="Shared Global Tooltip">
          <button>Minimal 1</button>
        </TooltipWrapper>
        <Tooltip>
          <button>button</button>
        </Tooltip>
      </TooltipProvider>,
    )
    const anchorElement = screen.getByText('Minimal 1')

    await userEvent.hover(anchorElement)

    let tooltip = null

    await waitFor(() => {
      tooltip = screen.getByRole('tooltip')
    })

    expect(tooltip).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
})
