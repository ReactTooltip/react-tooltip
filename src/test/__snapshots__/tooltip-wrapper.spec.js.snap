// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tooltip props basic tooltip component 1`] = `
<div>
  <span
    class="react-tooltip-wrapper"
    data-tooltip-content="Shared Global Tooltip"
    data-tooltip-place="bottom"
  >
    <button>
      Minimal 1
    </button>
  </span>
  <div
    class="react-tooltip"
    role="tooltip"
    style="left: 5px; top: 10px;"
  >
    <button>
      button
    </button>
    <div
      class="react-tooltip-arrow"
      style="left: 5px; top: -4px;"
    />
  </div>
</div>
`;
